### PR TITLE
[api] Reduce queries for radius_settings #195

### LINKED ISSUE #195: [api] Use JOIN when fetching organization in DispatchOrgMixin
Once https://github.com/openwisp/openwisp-radius/pull/190 is merged, we'll be accessing the organization radius settings object very often, which will generate at least one extra query each time.

Here's what we should do avoid that:

- [ ] add [assertNumQueries](https://docs.djangoproject.com/en/3.1/topics/testing/tools/#django.test.TransactionTestCase.assertNumQueries) in each API call executed in tests that were added in https://github.com/openwisp/openwisp-radius/pull/190
- [ ] change DispatchOrgMixin https://github.com/openwisp/openwisp-radius/blob/5acaad70db27165dd8c51f1ad65a28988fcfcdea/openwisp_radius/api/views.py#L419-L422
we need to forgo using the [get_obejct_or_404](https://docs.djangoproject.com/en/3.1/topics/http/shortcuts/#get-object-or-404) shortcut (resorting to the extended form mentioned in the docs instead), and we'll have to execute the query using `Organization.objects.select_related('radius_settings').get(slug=kwargs['slug'])`
- [ ] run the tests again, we expect the `assertNumQueries` we added in the first step to fail, now we should have at least one less executed query for each test

### PR DESCRIPTION
**Pull request by Purhan**
Closes #195

### CODE REVIEW

File: openwisp_radius/tests/test_api.py [hunk 1]
```diff
@@ -2594,16 +2594,17 @@ def test_register_201_phone_number_empty(self):
...
-            url,
-            {
-                'username': self._test_email,
-                'email': self._test_email,
-                'password1': 'password',
-                'password2': 'password',
-                'phone_number': phone_number,
-            },
-        )
+        with self.assertNumQueries(50):
```
Comment (nemesisdesign) [inline openwisp_radius/tests/test_api.py:2597]: :scream:

File: openwisp_radius/api/views.py [hunk 2]
```diff
@@ -419,7 +419,12 @@ def post(self, request, *args, **kwargs):
 
 class DispatchOrgMixin(object):
     def dispatch(self, *args, **kwargs):
-        self.organization = get_object_or_404(Organization, slug=kwargs['slug'])
+        try:
+            self.organization = Organization.objects.select_related(
+                'radius_settings'
+            ).get(slug=kwargs['slug'])
+        except Organization.DoesNotExist:
```
Comment (nemesisdesign) [inline openwisp_radius/api/views.py:426]: If we use the [DRF NotFound exception](https://www.django-rest-framework.org/api-guide/exceptions/#notfound), the API response should be more consistent with the rest of the responses from the API. Can you try it and let me know?

**Review by nemesisdesign — COMMENTED** [review]
Great @Purhan, looks good! See my comment below.

PS: please also rebase on the current master.

File: openwisp_radius/tests/test_api.py (hunk 1 above)
Comment (Purhan) [reply openwisp_radius/tests/test_api.py:2597]: :laughing: I was also freaked out when I saw this. Is there something we can do to reduce this?

File: openwisp_radius/api/views.py (hunk 2 above)
Comment (Purhan) [reply openwisp_radius/api/views.py:426]: For some reason, this is not being handled by the DRF exception handler. Look at the screenshot attached below, this is what I get.

image

File: openwisp_radius/tests/test_api.py (hunk 1 above)
Comment (Purhan) [reply openwisp_radius/tests/test_api.py:2597]: Also looks like Travis checks failed for this on some versions of django. This needs a different approach altogether...

**Review by nemesisdesign — COMMENTED** [review]
The failure on django 2.2 may be handled with an if which checks if the django version is less than 3 and in that case it would use 53 instead of 50 (adding also a TODO comment hinting to remove when django 2.2 support is removed).

File: openwisp_radius/tests/test_api.py [hunk 3]
```diff
@@ -2594,16 +2595,21 @@ def test_register_201_phone_number_empty(self):
...
-                'username': self._test_email,
-                'email': self._test_email,
-                'password1': 'password',
-                'password2': 'password',
-                'phone_number': phone_number,
-            },
-        )
+        numQueries = 50
+        # TODO: Remove this when support for django 2 is dropped
+        if django.get_version().startswith('2'):
```
Comment (nemesisdesign) [inline openwisp_radius/tests/test_api.py:2600]: this is an unusual way to do the comparison, I suggest to keep consistency with the way it's usually done, eg:
https://github.com/openwisp/openwisp-users/blob/fabef8d63912c0c492248079e82fb3092f9b8ab8/openwisp_users/base/models.py#L166

There are many ways of doing it, but that's the one of the simplest and the one we've been using in several ow modules.

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
Thanks @Purhan, only one minor issue below.

File: openwisp_radius/tests/test_api.py (hunk 3 above)
Comment (Purhan) [reply openwisp_radius/tests/test_api.py]: Done :+1:

File: openwisp_radius/api/views.py [hunk 4]
```diff
@@ -419,7 +419,12 @@ def post(self, request, *args, **kwargs):
...
+            raise Http404("No Organization matches the given query.")
```
Comment (nemesisdesign) [inline openwisp_radius/api/views.py:427]: spotted one minor issue, please use single quotes for consistency

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
Tested and it's good, there's one minor issue below I want you to fix. I could fix it myself but if you fix it we'll avoid it next time.

File: openwisp_radius/api/views.py (hunk 4 above)
Comment (Purhan) [reply openwisp_radius/api/views.py]: Done

**Review by nemesisdesign — APPROVED** [review]

**Pull request merged**